naming: Update metric and label name restrictions and recommendations with the latest context (#2626)
prometheus/docs: content/docs/practices/naming.md

    (exported by many client libraries)
  * <code><b>http</b>\_request\_duration\_seconds</code>
    (for all HTTP requests)
* ...MUST have a single unit (i.e. do not mix seconds with milliseconds, or seconds with bytes).
* ...SHOULD use base units (e.g. seconds, bytes, meters - not milliseconds, megabytes, kilometers). See [below](#base-units) for a list of base units.

> * ...SHOULD have a suffix describing the unit, in plural form. Note that an accumulating count has `total` as a suffix, in addition to the unit if applicable. Also note that this applies to units in the narrow sense (like the units in the table below), but not to countable things in general. For example, <code>connections</code> or <code>notifications</code> are not considered units for this rule and do not have to be at the end of the metric name. (See also examples in the next paragraph.)

Reviewer: I find the "units ... countable" part hard to follow.  As far as I can see there are no examples with `connections` or `notifications`, so referring to examples is perverse.  
Is this emphasis because we don't put `requests` in the UNIT metadata?

Given the long section you have added below, maybe remove this part and refer to that section?

Reviewer 2: These words have been added as a reaction to misunderstandings in the past. We can move it elsewhere, but it should stay prominent enough to avoid repeating the misunderstanding.

There _is_ an example below: 

```
* <code>http\_requests\_<b>total</b></code>
    (for a unit-less accumulating count)
```

Author: "narrow sense units" (: 

While odd it makes sense for names and it's something we recommend for years. I will keep this form for now, we can change later.

capacity of various queues in one metric is good, while mixing the capacity of a
queue with the current number of elements in the queue is not.

> ### Why including unit (and _total for counter) suffixes in metric names?

Reviewer: ```suggestion
### Why include unit (and _total for counter) suffixes in metric names?
```

Reviewer 2: Maybe even

```suggestion
### Why include unit and type suffixes in metric names?
```

And then later explain that there is `_total` for counters and `_info` for info metrics.
I would also suggest to put a literal suffix with the `_` always in back quotes for better readability.

Some metric naming conventions (e.g. OpenTelemetry) do not recommend or even do not allow
including information about a metric unit and type in the metric name. A common
argument is that those pieces of information are already defined somewhere else (e.g. schema, 
metadata, other labels, etc.).

Prometheus strongly recommends including unit and type in a metric name, even if you store that
information elsewhere, because of the two main reasons:

* **Metric consumption reliability and UX**: When interacting with a modern UI to 
use such a metric in PromQL, it's possible to display rich information about the metric's type and unit
(autocompletion, overlays, pop-ups). Unfortunately, interactive, adhoc querying in a powerful UI is only

> a tiny fraction of the metric consumption use journeys (CUJs). Metric consumption ecosystem is vast. Majority

Reviewer 2: Metric consumption ecosystem is vast. → The metric consumption ecosystem is vast.

of the consumption comes in a form of the plain YAML configuration for variety of observability tools like
alerting, recording, autoscaling, dashboard, analysis, processing, etc. It's **critical**, especially 

> during monitoring/SRE incident practices to look on PromQL expressions in the plain YAML and understand

Reviewer 2: ```suggestion
during monitoring/SRE incident practices to look on PromQL expressions in plain YAML and understand
```

But I wouldn't even restrict this to YAML. You might encounter the naked metric name outside of YAML, too.

the underlying metric type and unit you work with.
* **Metric collisions**: With the enormous adoption and metric changes over time, there are cases where lack
of unit and type information in the metric name will cause certain series to collide (e.g. `process_cpu` for seconds and milliseconds).

> NOTE: While Prometheus community experiments with ways to reduce the above negative effects

Author: True. However I don't want to give impression there is a solution planned that will make Prometheus to NOT recommend type and unit in names.

Maybe to NOT give this impression and NOT go into details I can skip this NOTE completely (:
